File: csharp/DCbor/DCbor/Walk.cs
namespace BlockchainCommons.DCbor;

/// <summary>
/// An element or element pair visited during CBOR tree traversal.
/// </summary>
public abstract class WalkElement
{
    private WalkElement() { }

    /// <summary>A single CBOR element.</summary>
    public sealed class SingleElement : WalkElement
    {
        public Cbor Value { get; }
        public SingleElement(Cbor value) { Value = value; }
    }

    /// <summary>A key-value pair from a map.</summary>
    public sealed class KeyValueElement : WalkElement
    {
        public Cbor Key { get; }
        public Cbor Value { get; }
        public KeyValueElement(Cbor key, Cbor value) { Key = key; Value = value; }
    }

    public Cbor? AsSingle() => this is SingleElement s ? s.Value : null;

    public (Cbor Key, Cbor Value)? AsKeyValue() =>
        this is KeyValueElement kv ? (kv.Key, kv.Value) : null;

    public string DiagnosticFlat()
    {
        return this switch
        {
            SingleElement s => s.Value.DiagnosticFlat(),
            KeyValueElement kv =>
                $"{kv.Key.DiagnosticFlat()}: {kv.Value.DiagnosticFlat()}",
            _ => throw new InvalidOperationException(),
        };
    }
}

/// <summary>
/// The type of incoming edge during traversal.
/// </summary>
public abstract class EdgeType : IEquatable<EdgeType>
{
    private EdgeType() { }

    public sealed class NoneEdge : EdgeType
    {
        internal static readonly NoneEdge Instance = new();
    }

    public sealed class ArrayElementEdge : EdgeType
    {
        public int Index { get; }
        public ArrayElementEdge(int index) { Index = index; }
    }

    public sealed class MapKeyValueEdge : EdgeType
    {
        internal static readonly MapKeyValueEdge Instance = new();
    }

    public sealed class MapKeyEdge : EdgeType
    {
        internal static readonly MapKeyEdge Instance = new();
    }

    public sealed class MapValueEdge : EdgeType
    {
        internal static readonly MapValueEdge Instance = new();
    }

    public sealed class TaggedContentEdge : EdgeType
    {
        internal static readonly TaggedContentEdge Instance = new();
    }

    // --- Factory ---

    public static EdgeType None => NoneEdge.Instance;
    public static EdgeType ArrayElement(int index) => new ArrayElementEdge(index);
    public static EdgeType MapKeyValue => MapKeyValueEdge.Instance;
    public static EdgeType MapKey => MapKeyEdge.Instance;
    public static EdgeType MapValue => MapValueEdge.Instance;
    public static EdgeType TaggedContent => TaggedContentEdge.Instance;

    // --- Label ---

    public string? Label()
    {
        return this switch
        {
            ArrayElementEdge a => $"arr[{a.Index}]",
            MapKeyValueEdge => "kv",
            MapKeyEdge => "key",
            MapValueEdge => "val",
            TaggedContentEdge => "content",
            NoneEdge => null,
            _ => null,
        };
    }

    // --- Equality ---

    public bool Equals(EdgeType? other)
    {
        if (other is null) return false;
        return (this, other) switch
        {
            (NoneEdge, NoneEdge) => true,
            (ArrayElementEdge a, ArrayElementEdge b) => a.Index == b.Index,
            (MapKeyValueEdge, MapKeyValueEdge) => true,
            (MapKeyEdge, MapKeyEdge) => true,
            (MapValueEdge, MapValueEdge) => true,
            (TaggedContentEdge, TaggedContentEdge) => true,
            _ => false,
        };
    }

    public override bool Equals(object? obj) => Equals(obj as EdgeType);

    public override int GetHashCode()
    {
        return this switch
        {
            NoneEdge => 0,
            ArrayElementEdge a => HashCode.Combine(1, a.Index),
            MapKeyValueEdge => 2,
            MapKeyEdge => 3,
            MapValueEdge => 4,
            TaggedContentEdge => 5,
            _ => -1,
        };
    }
}

/// <summary>
/// A visitor function called for each element during CBOR tree traversal.
/// Returns (newState, stop) where stop=true prevents descent into children.
/// </summary>
public delegate (TState State, bool Stop) CborVisitor<TState>(
    WalkElement element, int level, EdgeType edge, TState state);

/// <summary>
/// Walk extension methods for CBOR traversal.
/// </summary>
public static class CborWalk
{
    /// <summary>
    /// Walks the CBOR structure, calling the visitor for each element.
    /// </summary>
    public static void Walk<TState>(this Cbor cbor, TState initialState, CborVisitor<TState> visitor)
    {
        WalkInternal(cbor, 0, EdgeType.None, initialState, visitor);
    }

    private static void WalkInternal<TState>(
        Cbor cbor, int level, EdgeType incomingEdge, TState state, CborVisitor<TState> visitor)
    {
        // Visit this element as a single element
        var element = new WalkElement.SingleElement(cbor);
        var (newState, stop) = visitor(element, level, incomingEdge, state);
        if (stop) return;

        int nextLevel = level + 1;
        switch (cbor.Case)
        {
            case CborCase.ArrayCase a:
                for (int i = 0; i < a.Value.Count; i++)
                {
                    WalkInternal(a.Value[i], nextLevel, EdgeType.ArrayElement(i), newState, visitor);
                }
                break;

            case CborCase.MapCase m:
                foreach (var (key, value) in m.Value)
                {
                    // First visit the key-value pair as a semantic unit
                    var kvElement = new WalkElement.KeyValueElement(key, value);
                    var (kvState, kvStop) = visitor(kvElement, nextLevel, EdgeType.MapKeyValue, newState);
                    if (kvStop) continue;

                    // Then visit key and value individually
                    WalkInternal(key, nextLevel, EdgeType.MapKey, kvState, visitor);
                    WalkInternal(value, nextLevel, EdgeType.MapValue, kvState, visitor);
                }
                break;

            case CborCase.TaggedCase tg:
                WalkInternal(tg.Item, nextLevel, EdgeType.TaggedContent, newState, visitor);
                break;

            // Primitives have no children
            default:
                break;
        }
    }
}

File: python/dcbor/src/dcbor/walk.py
from __future__ import annotations

from collections.abc import Callable
from enum import Enum, auto
from typing import TYPE_CHECKING, TypeVar

if TYPE_CHECKING:
    from .cbor import CBOR

S = TypeVar("S")


class WalkElement:
    __slots__ = ("_single", "_key", "_value")

    def __init__(
        self,
        single: CBOR | None = None,
        key: CBOR | None = None,
        value: CBOR | None = None,
    ) -> None:
        self._single = single
        self._key = key
        self._value = value

    @staticmethod
    def make_single(cbor: CBOR) -> WalkElement:
        return WalkElement(single=cbor)

    @staticmethod
    def make_key_value(key: CBOR, value: CBOR) -> WalkElement:
        return WalkElement(key=key, value=value)

    def as_single(self) -> CBOR | None:
        return self._single

    def as_key_value(self) -> tuple[CBOR, CBOR] | None:
        if self._key is not None and self._value is not None:
            return (self._key, self._value)
        return None

    def diagnostic_flat(self) -> str:
        if self._single is not None:
            return self._single.diagnostic_flat()
        if self._key is not None and self._value is not None:
            return f"{self._key.diagnostic_flat()}: {self._value.diagnostic_flat()}"
        return ""


class EdgeType(Enum):
    NONE = auto()
    ARRAY_ELEMENT = auto()
    MAP_KEY_VALUE = auto()
    MAP_KEY = auto()
    MAP_VALUE = auto()
    TAGGED_CONTENT = auto()

    def label(self, index: int | None = None) -> str | None:
        match self:
            case EdgeType.ARRAY_ELEMENT:
                return f"arr[{index}]"
            case EdgeType.MAP_KEY_VALUE:
                return "kv"
            case EdgeType.MAP_KEY:
                return "key"
            case EdgeType.MAP_VALUE:
                return "val"
            case EdgeType.TAGGED_CONTENT:
                return "content"
            case EdgeType.NONE:
                return None


Visitor = Callable[[WalkElement, int, EdgeType, S], tuple[S, bool]]


def walk_impl(cbor: CBOR, state: S, visitor: Visitor[S]) -> S:
    return _walk(cbor, 0, EdgeType.NONE, state, visitor)


def _walk(
    cbor: CBOR,
    level: int,
    edge: EdgeType,
    state: S,
    visitor: Visitor[S],
) -> S:
    from .cbor import CBORCase

    element = WalkElement.make_single(cbor)
    state, stop = visitor(element, level, edge, state)
    if stop:
        return state

    next_level = level + 1

    match cbor.case:
        case CBORCase.ARRAY:
            for child in cbor.value:
                state = _walk(child, next_level, EdgeType.ARRAY_ELEMENT, state, visitor)

        case CBORCase.MAP:
            for key, value in cbor.value.iter():
                kv_element = WalkElement.make_key_value(key, value)
                state, stop = visitor(kv_element, next_level, EdgeType.MAP_KEY_VALUE, state)
                if stop:
                    continue
                state = _walk(key, next_level, EdgeType.MAP_KEY, state, visitor)
                state = _walk(value, next_level, EdgeType.MAP_VALUE, state, visitor)

        case CBORCase.TAGGED:
            _, content = cbor.value
            state = _walk(content, next_level, EdgeType.TAGGED_CONTENT, state, visitor)

    return state
